File: daily_python/3394_Check_if_Grid_can_be_Cut_into_Sections.py
import os
import sys
from typing import List

class Solution:
    def checkValidCuts(self, n: int, rectangles: List[List[int]]) -> bool:
        r = len(rectangles)
        cut = 0
        end = 0
        sorted_horizotal = sorted(rectangles, key = lambda x : x[0])
        sorted_vertical = sorted(rectangles, key = lambda x : x[1])
        # cut vertical
        for startx, starty, endx, endy in sorted_horizotal:
            if startx >= end:
                cut += 1
                if cut == 3:
                    return True
            end = max(end, endx)
        cut = 0
        end = 0
        # cut horizontal
        for startx, starty, endx, endy in sorted_vertical:
            if starty >= end:
                cut += 1
                if cut == 3:
                    return True
            end = max(end, endy)
        return False
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            rectangles = [list(map(int, query.split(','))) for query in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((n, rectangles))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, rectangles in dataset:
        results.append(solution.checkValidCuts(n, rectangles))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/3394_Check_if_Grid_can_be_Cut_into_Sections.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool checkValidCuts(int n, vector<vector<int>>& rectangles) {
        auto cut_vertical = [](vector<int>& a, vector<int>& b){return a[0] < b[0];};
        auto cut_horizontal = [](vector<int>& a, vector<int>& b){return a[1] < b[1];};
        int cut = 0;
        int r = rectangles.size();
        int end = 0;
        sort(rectangles.begin(), rectangles.end(), cut_vertical);
        for (int i = 0; i < r; i++){
            if (rectangles[i][0] >= end) cut++;
            if (cut == 3) return true;
            end = max(end, rectangles[i][2]);
        }
        sort(rectangles.begin(), rectangles.end(), cut_horizontal);
        cut = 0;
        end = 0;
        for (int i = 0; i < r; i++){
            if (rectangles[i][1] >= end) cut++;
            if (cut == 3) return true;
            end = max(end, rectangles[i][3]);
        }
        return false;
    }
};

int main(){
    int n = 5;
    vector<vector<int>> rectangles = {{1,0,5,2},{0,2,2,4},{3,2,5,3},{0,4,4,5}};
    // int n = 4;
    // vector<vector<int>> rectangles = {{0,0,1,1},{2,0,3,4},{0,2,2,3},{3,0,4,3}};
    // int n = 4;
    // vector<vector<int>> rectangles = {{0,2,2,4},{1,0,3,2},{2,2,3,4},{3,0,4,2},{3,2,4,4}};
    Solution S;
    cout << S.checkValidCuts(n, rectangles) << endl;
    return 0;
}